### PR TITLE
Debugger: Source-level Debugging

### PR DESCRIPTION
**Pull request by itamar8910**
This PR adds the following capabilities to `sdb`:
- Print the current source location, if available.
- Add a breakpoint at a source location (`bp <file>:<line>`)
- `sl`  - step to the next line in source.

To do source-level debugging, we parse the DWARF `.debug_line` data to get line number information.

I tried adding `-g` to `CXXFLAGS` for all the the executables in the build process,
but since we don't currently have shared libraries the debug info for libraries is duplicated many times over, and we run out of space in the disk image :disappointed:

So for now, we can test this out with small programs.
An example:

`test.c`
```c
1. #include <stdio.h>
2.
3. int foo();
4.
5. int main()
6. {
7.    int a = foo();
8.    printf("%d\n", a);
9. }
```
`foo.c`
```c
1. int foo(){
2.    int x = 1;
3.    int y = 3;
4.    return x + y;
5. }
```
sdb-source2

### CODE REVIEW

File: Libraries/LibDebug/DebugInfo.h [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+#include <LibELF/Loader.h>
+#include <Libraries/LibDebug/Dwarf/LineProgram.h>
+
+class DebugInfo {
+public:
+    explicit DebugInfo(NonnullRefPtr<const ELF::Loader> elf);
+
+    struct SourcePosition {
+        String file_path;
+        size_t line_number;
```
Comment (awesomekling) [inline Libraries/LibDebug/DebugInfo.h:41]: Please explicitly initialize primitives like this.
```suggestion
        size_t line_number { 0 };
```

File: Libraries/LibDebug/Dwarf/LineProgram.h [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+#pragma once
+#include <AK/BufferStream.h>
+#include <AK/Vector.h>
+
+class LineProgram {
+public:
+    explicit LineProgram(BufferStream& stream);
+
+    struct LineInfo {
+        u32 address;
```
Comment (awesomekling) [inline Libraries/LibDebug/Dwarf/LineProgram.h:36]: Please explicitly initialize primitives.

File: Libraries/LibDebug/Dwarf/LineProgram.h [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+        String file;
+        size_t line;
```
Comment (awesomekling) [inline Libraries/LibDebug/Dwarf/LineProgram.h:38]: Please explicitly initialize primitives.

File: Libraries/LibDebug/Dwarf/LineProgram.h [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+    enum ExtendedOpcodes {
+        EndSequence = 1,
+        SetAddress,
+        DefineFile,
+        SetDiscriminator,
+    };
+
+    struct FileEntry {
+        String name;
+        size_t directory_index;
```
Comment (awesomekling) [inline Libraries/LibDebug/Dwarf/LineProgram.h:93]: Please explicitly initialize primitives.

File: Libraries/LibDebug/Dwarf/LineProgram.h [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+    size_t m_unit_offset { 0 };
+    UnitHeader32 m_unit_header {};
+    Vector<String> m_source_directories;
+    Vector<FileEntry> m_source_files;
+
+    // The registers of the "line program" virtual machine
+    u32 m_address { 0 };
+    size_t m_line { 0 };
+    size_t m_file_index { 0 };
+    bool m_is_statement = { false };
```
Comment (awesomekling) [inline Libraries/LibDebug/Dwarf/LineProgram.h:110]: This looks weird. Either `= false` or `{ false }` but let's not have both :D

File: Libraries/LibDebug/DebugSession.h [hunk 6]
```diff
@@ -104,14 +108,14 @@ class DebugSession {
...
-    static constexpr u8 BREAKPOINT_INSTRUCTION
-        = 0xcc;
+    static constexpr u8 BREAKPOINT_INSTRUCTION = 0xcc;
 
     int m_debugee_pid { -1 };
     bool m_is_debugee_dead { false };
 
     MappedFile m_executable;
-    ELF::Loader m_elf;
+    NonnullRefPtr<const ELF::Loader> m_elf;
```
Comment (awesomekling) [inline Libraries/LibDebug/DebugSession.h:117]: The fact that this didn't break immediately when you made `ELF::Loader` refcounted is why I like to add static construction helpers for ref-counted objects :)

**Comment by sunverwerth** [comment]
@itamar8910 @awesomekling Now somebody needs to implement controlling `sdb` from within HackStudio. 😎

**Pull request merged**